Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+
+At the same time it makes us to repeat all the configurations in each target which is a lot of boilerplate for a simple flag and these configs grew a lot over time. Or we have to use protocol extensions with default values which makes it trickier to understand the actual value for particular target as we need to make sure this default is not overriden anywhere else.
+
+### Local Feature Switches boilerplate
+
+It takes a lot of boilerplate to define a local feature switch - we need to change the code in several places (in one file though), need to manually add an entry to the `Root.plist` and make sure that it's in the right place and in the write file. We need to make sure we move it to the different section (for remote feature toggles) if we move the flag to the Firebase, otherwise we won't be able to override it in the Settings app and it will always have a default value if we override any other flag.
+
+
+### Debugging
+
+It's not clear what feature switches affect what screens. It's somtimes frustrating to go through a complex flow and then realise that the right feature switch was not turned on/off in the Settings app or local overrides were active and to go through the flow again (as we don't observe changes in the feature switches at runtime and require application to restart).
+
+
+## Proposed solution
+
+To address these issues we suggest the following changes in our feature switches architecture:
+
+1. Feature toggles and static configurations related to the particular feature should be defined in this feature module instead of shared module or in the app. This follows interface segregation principle and goes along with improving modularisation - one won't need to change common module with feature switches and recompile all of it (and potentially everything else) when they only need to recompile module that uses this feature switch. This way we also configure our frameworks the same way as we configure any other 3rd party framework and can use the similar infrastrauctore for that (i.e. dedicated instances responsibule for configuring each feature rather than doing that all in the app delegate).
+
+2. Replace `AppConfigurationProtocol` and all related protocols with structs. This will decrease amount of boilerplate we have to write when adding new application configuration. Application configuration will be only concerned with application level configurations which are not directly related to any feature module.
+
+3. All feature switches access points should be moved to the `Current` and consequentially should be only used by builders - view models, flow controllers and renderers should have them injected through their constructors.
+
+4. Root.plist should be generated based on the code that declares feature switches.
```

Comment:
You mean using [this](https://github.com/apple/swift-syntax) for generation? Again, not sure right now how that might work.

We could possibly go the other way and use SwiftGen to transform plists into code instead :) But I'd prefer code first tbh.

Reply:
Yes, we can use SwiftSyntax for that or just roll out our custom parser. Using SwiftGen is a nice suggestion as we will have it used in the project already anyway and will probable require less effort (only if it supports parsing custom plist resources though cc @AliSoftware). But again, editing plist is much less friendly experience than editing code.

Reply:
SwiftGen would be able to generate Swift code from arbitrary Plist (we have a PlistParser now so yes you can do this even for custom plist resources).
But if we want to do the opposite and generate Plist from code, we'd need to use either Sourcery or SwiftSyntax (at that point introducing Sourcery would make more sense to open ourselves to other uses of it in other situations in the future too while SwiftSyntax would require custom coding that would only work for this specific task)